### PR TITLE
rpc: improve the realtime notify performance by 30%

### PR DESCRIPTION
**Pull request by jsvisa**
I'm running 10 websocket clients(mainly to test #27439) to receive the real-time events, it turns out the `rpc.Notify` eat too much CPU, see the profile file below:

image

And found there are too many CPUs used in JSON.marshaling, and after digging in the source code, seems we can improve the performance by reducing 1 JSON.marshaling. So I add a benchmark, here is the result of the benchmark: 

```bash
go test -bench=.  -v -run BenchmarkNotify -count 10
```

> old 

```txt
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/rpc
cpu: Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
BenchmarkNotify
BenchmarkNotify-16         19371             66857 ns/op
BenchmarkNotify-16         19032             59029 ns/op
BenchmarkNotify-16         19263             67036 ns/op
BenchmarkNotify-16         16597             61500 ns/op
BenchmarkNotify-16         19549             65228 ns/op
BenchmarkNotify-16         16611             72045 ns/op
BenchmarkNotify-16         16962             68970 ns/op
BenchmarkNotify-16         16431             66456 ns/op
BenchmarkNotify-16         23557             49392 ns/op
BenchmarkNotify-16         23340             53930 ns/op
PASS
ok      github.com/ethereum/go-ethereum/rpc     18.224s
```

> new

```txt
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/rpc
cpu: Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
BenchmarkNotify
BenchmarkNotify-16         24076             45048 ns/op
BenchmarkNotify-16         25297             42735 ns/op
BenchmarkNotify-16         23407             45985 ns/op
BenchmarkNotify-16         34939             46708 ns/op
BenchmarkNotify-16         36223             45710 ns/op
BenchmarkNotify-16         24590             46140 ns/op
BenchmarkNotify-16         22858             45640 ns/op
BenchmarkNotify-16         26013             47350 ns/op
BenchmarkNotify-16         31420             44238 ns/op
BenchmarkNotify-16         26156             45340 ns/op
PASS
ok      github.com/ethereum/go-ethereum/rpc     17.150s
```

generate the bench differ:

```bash
benchstat old.txt new.txt
old.txt:5: missing iteration count
new.txt:5: missing iteration count
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/rpc
cpu: Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
          │   old.txt    │               new.txt               │
          │    sec/op    │   sec/op     vs base                │
Notify-16   65.84µ ± 18%   45.68µ ± 3%  -30.63% (p=0.000 n=10)

```

### CODE REVIEW

File: rpc/subscription.go [hunk 1]
```diff
@@ -181,8 +176,11 @@ func (n *Notifier) activate() error {
 	return nil
 }
 
-func (n *Notifier) send(sub *Subscription, data json.RawMessage) error {
-	params, _ := json.Marshal(&subscriptionResult{ID: string(sub.ID), Result: data})
+func (n *Notifier) send(sub *Subscription, data interface{}) error {
+	params, _ := json.Marshal(struct {
+		ID     string      `json:"subscription"`
+		Result interface{} `json:"result,omitempty"`
+	}{ID: string(sub.ID), Result: data})
```
Comment (holiman) [inline rpc/subscription.go:183]: Why not just modify `subscriptionResult`?

Comment (jsvisa) [reply rpc/subscription.go:183]: `subscriptionResult` this struct is also used by the websocket client, modify that struct may affect the client's marshal/unmarshal logic. Let me see if we can change it together

**Comment by holiman** [comment]
The percentages weren't quite right: you just discarded the output, but in reality it will be json-encoded at least one more time, which adds to the overhead (and makes the impact less). 

Aside from that, I noticed that we could theoretically take this idea even further, like this: 

```
diff --git a/rpc/subscription.go b/rpc/subscription.go
index 947767e963..920ad960ce 100644
--- a/rpc/subscription.go
+++ b/rpc/subscription.go
@@ -177,16 +177,24 @@ func (n *Notifier) activate() error {
 }
 
 func (n *Notifier) send(sub *Subscription, data interface{}) error {
-	params, _ := json.Marshal(struct {
+	thing := struct {
 		ID     string      `json:"subscription"`
 		Result interface{} `json:"result,omitempty"`
-	}{ID: string(sub.ID), Result: data})
+	}{ID: string(sub.ID), Result: data}
+
 	ctx := context.Background()
 
-	msg := &jsonrpcMessage{
+	msg := struct {
+		Version string          `json:"jsonrpc,omitempty"`
+		ID      json.RawMessage `json:"id,omitempty"`
+		Method  string          `json:"method,omitempty"`
+		Params  interface{}     `json:"params,omitempty"`
+		Error   *jsonError      `json:"error,omitempty"`
+		Result  json.RawMessage `json:"result,omitempty"`
+	}{
 		Version: vsn,
 		Method:  n.namespace + notificationMethodSuffix,
-		Params:  params,
+		Params:  thing,
 	}
 	return n.h.conn.writeJSON(ctx, msg, false)
 }
```
That is, always avoid putting `json.RawMessage` into structs, and instead keep things in object-form as long as possible in the encoding pipeline. 

I don't know if this is worth doing, but if it is, we should probably do it generally, not just in this particular instance.

**Comment by jsvisa** [comment]
@holiman thanks for your tips. After using the new benchmark code, there are also 23% improvement.

```
>$ cat old.txt
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/rpc
cpu: Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
BenchmarkNotify
BenchmarkNotify-16         17034             87875 ns/op
BenchmarkNotify-16         14574             80534 ns/op
BenchmarkNotify-16         14750             95860 ns/op
BenchmarkNotify-16         13078             90346 ns/op
BenchmarkNotify-16         14583             74846 ns/op
BenchmarkNotify-16         12854             87367 ns/op
BenchmarkNotify-16         14354             82897 ns/op
BenchmarkNotify-16         14950             76899 ns/op
BenchmarkNotify-16         13184             89349 ns/op
BenchmarkNotify-16         15166             77653 ns/op
PASS
ok      github.com/ethereum/go-ethereum/rpc     21.838s
>$ cat new.txt
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/rpc
cpu: Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
BenchmarkNotify
BenchmarkNotify-16         17634             59332 ns/op
BenchmarkNotify-16         19249             65428 ns/op
BenchmarkNotify-16         18812             71263 ns/op
BenchmarkNotify-16         17366             66708 ns/op
BenchmarkNotify-16         17824             69048 ns/op
BenchmarkNotify-16         17083             64215 ns/op
BenchmarkNotify-16         19155             61733 ns/op
BenchmarkNotify-16         21247             66564 ns/op
BenchmarkNotify-16         22184             63637 ns/op
BenchmarkNotify-16         20374             56697 ns/op
PASS
ok      github.com/ethereum/go-ethereum/rpc     18.847s

>$ benchstat old.txt new.txt
old.txt:5: missing iteration count
new.txt:5: missing iteration count
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/rpc
cpu: Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
          ?   old.txt    ?               new.txt               ?
          ?    sec/op    ?   sec/op     vs base                ?
Notify-16   85.13µ ± 10%   64.82µ ± 8%  -23.86% (p=0.000 n=10)

```

**Comment by jsvisa** [comment]
> That is, always avoid putting json.RawMessage into structs, and instead keep things in object-form as long as possible in the encoding pipeline.

Try to adapt more to that style, and see if the perform differs

**Comment by holiman** [comment]
```
[user@work go-ethereum]$ benchstat after_0.txt after_1.txt 
name      old time/op  new time/op  delta
Notify-8  53.0µs ±17%  40.8µs ±14%  -23.08%  (p=0.000 n=10+10)
[user@work go-ethereum]$ benchstat after_1.txt after_2.txt 
name      old time/op  new time/op  delta
Notify-8  40.8µs ±14%  27.1µs ± 4%  -33.54%  (p=0.000 n=10+8)
```

I'm getting first a 23% improv, then another 33%. Not suprisingly

File: rpc/subscription_test.go [hunk 2]
```diff
@@ -218,3 +225,56 @@ func readAndValidateMessage(in *json.Decoder) (*subConfirmation, *subscriptionRe
...
+		h:         &handler{conn: &mockConn{json.NewEncoder(out)}},
+		sub:       &Subscription{ID: id},
+		activated: true,
+	}
+	msg := &types.Header{
+		ParentHash: common.HexToHash("0x01"),
+		Number:     big.NewInt(100),
+	}
+	notifier.Notify(id, msg)
+	have := string(out.Bytes())
```
Comment (holiman) [inline rpc/subscription_test.go:275]:
```suggestion
	have := strings.TrimSpace(string(out.Bytes()))
```

Comment (holiman) [reply rpc/subscription_test.go:275]: My bad, checked in a faulty test

**Comment by jsvisa** [comment]
@holiman I think we can declare different JSON-RPC structs for the request and response, using `json.RawMessage` for requests, and `interface` for response, wdyt?

**Comment by holiman** [comment]
Yes, might be a good solution. But this really is @fjl's domains, so don't take my idea as a statement of "this is how we want to do it".

**Comment by jsvisa** [comment]
OK, I'll pause here, and wait for @fjl's opinion

**Comment by jsvisa** [comment]
Here is a simple script to benchmark the single vs multi `json.Marshal`:


```go
package main

import (
	"encoding/json"
	"testing"
)

type rawResult struct {
	ID     string          `json:"subscription"`
	Result json.RawMessage `json:"result,omitempty"`
}

type rawMessage struct {
	Version string          `json:"jsonrpc,omitempty"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method,omitempty"`
	Params  json.RawMessage `json:"params,omitempty"`
	Result  json.RawMessage `json:"result,omitempty"`
}

type interfaceResult struct {
	ID     string      `json:"subscription"`
	Result interface{} `json:"result,omitempty"`
}

type interfaceMessage struct {
	Version string          `json:"jsonrpc,omitempty"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method,omitempty"`
	Params  interface{}     `json:"params,omitempty"`
	Result  interface{}     `json:"result,omitempty"`
}

type Header struct {
	Number int
}

func BenchmarkInterface(b *testing.B) {
	header := Header{Number: 1}
	for i := 0; i < b.N; i++ {
		params := interfaceResult{ID: "id", Result: header}
		msg := interfaceMessage{
			Version: "2.0",
			Method:  "rpc_test",
			Params:  params,
		}
		json.Marshal(msg)
	}
}

func BenchmarkRaw(b *testing.B) {
	header := Header{Number: 1}
	for i := 0; i < b.N; i++ {
		enc, _ := json.Marshal(header)
		params, _ := json.Marshal(&rawResult{ID: "id", Result: enc})
		msg := rawMessage{
			Version: "2.0",
			Method:  "rpc_test",
			Params:  params,
		}
		json.Marshal(msg)
	}
}
```

And here is a result in my PC:

```bash
go test -bench=. -v
goos: linux
goarch: amd64
pkg: github.com/jsvisa/gochain
cpu: Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
BenchmarkInterface
BenchmarkInterface-16             682720              1941 ns/op
BenchmarkRaw
BenchmarkRaw-16                   375933              3832 ns/op
PASS
ok      github.com/jsvisa/gochain       4.670s
```

**Comment by s1na** [comment]
This seems to resonate with https://github.com/golang/go/issues/33422. Unrelated: but there I saw a proposal for `encoding/json/v2` https://github.com/golang/go/discussions/63397. Looking cool!

**Comment by s1na** [comment]
I'm curious how the performance will be if the parameter passed to `Notify` is not a pointer, e.g. `h types.Header`. I imagine there will be some extra copying happening. That said our current subscriptions pass mostly pointers, with the exceptions of the pending transaction subscription which passes in a `common.Hash`.

File: rpc/subscription_test.go [hunk 3]
```diff
@@ -17,12 +17,19 @@
...
+	"math/big"
 	"net"
 	"strings"
 	"testing"
 	"time"
+
+	"bytes"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"io"
```
Comment (holiman) [inline rpc/subscription_test.go:32]: Please fix imports

**Review by holiman — APPROVED** [review]
LGTM aside from import-nits

File: rpc/subscription_test.go (hunk 3 above)
Comment (jsvisa) [reply rpc/subscription_test.go]: Didn't notice that, fixed.

File: rpc/subscription_test.go [hunk 4]
```diff
@@ -218,3 +225,56 @@ func readAndValidateMessage(in *json.Decoder) (*subConfirmation, *subscriptionRe
...
+		h:         &handler{conn: &mockConn{json.NewEncoder(out)}},
+		sub:       &Subscription{ID: id},
+		activated: true,
+	}
+	msg := &types.Header{
+		ParentHash: common.HexToHash("0x01"),
+		Number:     big.NewInt(100),
+	}
+	notifier.Notify(id, msg)
+	have := strings.TrimSpace(string(out.Bytes()))
```
Comment (holiman) [inline rpc/subscription_test.go:275]:
```suggestion
	have := strings.TrimSpace(out.String())
```

**Pull request merged**